aspects/aspects.go

```diff
@@ -111,46 +112,121 @@ func NewAspectDirectory(name string, aspects map[string]interface{}, dataBag Dat
+		accessPatterns: make([]*accessPattern, 0, len(aspectPatterns)),
+		directory:      dir,
+	}
 
-			aspect.accessPatterns = append(aspect.accessPatterns, accPattern)
+	for _, aspectPattern := range aspectPatterns {
+		name, ok := aspectPattern["name"]
+		if !ok || name == "" {
+			return nil, errors.New(`access patterns must have a "name" field`)
 		}
 
-		aspectDir.aspects[name] = aspect
+		path, ok := aspectPattern["path"]
+		if !ok || path == "" {
+			return nil, errors.New(`access patterns must have a "path" field`)
+		}
+
+		if err := validateNamePathPair(name, path); err != nil {
+			return nil, err
+		}
+
+		accPattern, err := newAccessPattern(name, path, aspectPattern["access"])
+		if err != nil {
+			return nil, err
+		}
+
+		aspect.accessPatterns = append(aspect.accessPatterns, accPattern)
+	}
+
+	return aspect, nil
+}
+
+// validateNamePathPair checks that:
+//     * names and paths are composed of valid subkeys (see: validateAspectString)
+//     * all placeholders in a name are in the path and vice-versa
+func validateNamePathPair(name, path string) error {
+	if err := validateAspectDottedPath(name); err != nil {
+		return fmt.Errorf("invalid access name %q: %w", name, err)
+	}
+
+	if err := validateAspectDottedPath(path); err != nil {
+		return fmt.Errorf("invalid path %q: %w", path, err)
+	}
+
+	namePlaceholders, pathPlaceholders := getPlaceholders(name), getPlaceholders(path)
+	if len(namePlaceholders) != len(pathPlaceholders) {
+		return fmt.Errorf("access name %q and path %q have mismatched placeholders", name, path)
+	}
+
+	for placeholder := range namePlaceholders {
+		if !pathPlaceholders[placeholder] {
+			return fmt.Errorf("placeholder %q from access name %q is absent from path %q",
+				placeholder, name, path)
+		}
+	}
+
+	return nil
+}
+
+var (
+	subkeyRegex      = "(?:[a-z0-9]+-?)*[a-z](?:-?[a-z0-9])*"
+	validSubkey      = regexp.MustCompile(fmt.Sprintf("^%s$", subkeyRegex))
+	validPlaceholder = regexp.MustCompile(fmt.Sprintf("^{%s}$", subkeyRegex))
+)
+
+// validateAspectDottedPath validates that names/paths in an aspect definition are:
+//     * composed of non-empty, dot-separated subkeys with optional placeholders ("foo.{bar}")
+//     * non-placeholder subkeys are made up of lowercase alphanumeric ASCII characters,
+//			optionally with dashes between alphanumeric characters (e.g., "a-b-c")
+//     * placeholder subkeys are composed of non-placeholder subkeys wrapped in curly brackets
+func validateAspectDottedPath(path string) (err error) {
+	subkeys := strings.Split(path, ".")
+
+	for _, subkey := range subkeys {
+		if subkey == "" {
+			return errors.New("cannot have empty subkeys")
+		}
+
+		if !(validSubkey.MatchString(subkey) || validPlaceholder.MatchString(subkey)) {
+			return fmt.Errorf("invalid subkey %q", subkey)
+		}
+	}
+
+	return nil
+}
+
+// getPlaceholders returns the set of placeholders in the string or nil, if
+// there is none.
+func getPlaceholders(aspectStr string) map[string]bool {
```

Comment:
There are some similarities between this function and the code in https://github.com/snapcore/snapd/blob/425231ec209a38cb0bb54ada73bb3db0d8f59fc6/aspects/aspects.go#L239-L263, I wonder if the common parts can be refactored?

Reply:
I tried to refactor it at the time but the casts to the interface implementations make it a bit difficult/not worth it